source/token_provider_sso_profile.c

```diff
@@ -0,0 +1,184 @@
+    profile_name = aws_get_profile_name(allocator, &profile_name_override);
+    if (!profile_name) {
+        AWS_LOGF_ERROR(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "sso-profile: token parser failed to resolve profile name");
+        aws_raise_error(AWS_AUTH_SSO_TOKEN_PROVIDER_SOURCE_FAILURE);
+        goto cleanup;
+    }
+    config_collection = aws_load_profile_collection_from_config_file(allocator, config_file_name_override);
+
+    if (!config_collection) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "sso-profile: token parser could not load or parse"
+            " a config file.");
+        aws_raise_error(AWS_AUTH_SSO_TOKEN_PROVIDER_SOURCE_FAILURE);
+        goto cleanup;
+    }
+
+    const struct aws_profile *profile = aws_profile_collection_get_profile(config_collection, profile_name);
+
+    if (!profile) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "sso-profile: token provider could not load"
+            " a profile at %s.",
+            aws_string_c_str(profile_name));
+        aws_raise_error(AWS_AUTH_SSO_TOKEN_PROVIDER_SOURCE_FAILURE);
+        goto cleanup;
+    }
+
+    const struct aws_profile_property *sso_start_url_property =
+        aws_profile_get_property(profile, s_profile_sso_start_url_name);
+
+    if (!sso_start_url_property) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "sso-profile: token parser failed to find sso_start_url in profile");
+        aws_raise_error(AWS_AUTH_SSO_TOKEN_PROVIDER_SOURCE_FAILURE);
+        goto cleanup;
+    }
+
+    token_path = aws_construct_token_path(allocator, aws_profile_property_get_value(sso_start_url_property));
+    if (!token_path) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "sso-profile: token parser failed to construct token path in profile");
+        goto cleanup;
+    }
+
+cleanup:
+    aws_string_destroy(profile_name);
+    aws_profile_collection_release(config_collection);
+    return token_path;
+}
+
+struct aws_credentials_provider *aws_token_provider_new_sso_profile(
+    struct aws_allocator *allocator,
+    const struct aws_token_provider_sso_profile_options *options) {
+    struct aws_string *token_path =
+        s_construct_profile_token_path(allocator, options->profile_name_override, options->config_file_name_override);
+    if (!token_path) {
+        return NULL;
+    }
+    struct aws_credentials_provider *provider = NULL;
+    struct aws_token_provider_profile_impl *impl = NULL;
+
+    aws_mem_acquire_many(
+        allocator,
+        2,
+        &provider,
+        sizeof(struct aws_credentials_provider),
+        &impl,
+        sizeof(struct aws_token_provider_profile_impl));
+    AWS_ZERO_STRUCT(*provider);
+    AWS_ZERO_STRUCT(*impl);
+    aws_credentials_provider_init_base(provider, allocator, &s_aws_token_provider_profile_vtable, impl);
+    impl->token_file_path = aws_string_new_from_string(allocator, token_path);
```

Comment:
Trivial: we can avoid this copy by allocating *impl early, just need to release when the token_path failed to construct.

Reply:
This is debatable because SSOCredentialsProvider will be added to default chain. Here are the three options that I could think of.

1. Allocate impl early and release it if token_path creation failed. As this will be part of default chain, It will be frequently visited path that we try to create SSOCredentialsProvider but the config is not right. This will result in unnecessary allocations.
2. We could simply pass the token by reference to the impl instead of making a deep copy and destroying it. This is the most optimal but suffers from readability as the allocation and deallocaiton are not on the same level.
3. Current approach which does an extra allocation for token path.